astropy/coordinates/sky_coordinate_parsers.py

```diff
@@ -588,7 +588,19 @@ def _get_representation_attrs(frame, units, kwargs):
     for frame_attr_name, repr_attr_class, unit in zip(frame_attr_names, repr_attr_classes, units):
         value = kwargs.pop(frame_attr_name, None)
         if value is not None:
-            valid_kwargs[frame_attr_name] = repr_attr_class(value, unit=unit)
+            try:
+                valid_kwargs[frame_attr_name] = repr_attr_class(value, unit=unit)
+            except u.UnitConversionError as err:
+                error_message = "Unit '{unit}' ({type}) could not be applied to '{name}'. ".format(
+                    unit=unit, type=unit.physical_type, name=frame_attr_name
+                )
+                error_message += (
+                    "This can occur when passing units for some coordinate components "
+                    "when other components are specified as Quantity objects. "
+                    "Either pass a list of units for all components (and unit-less coordinate data), "
+                    "or pass Quantities for all components"
+                )
+                raise ValueError(error_message) from err
```

Comment:
I'd stick with a `UnitConversionError` (which is a subclass of `ValueError`). You could also consider just adding to the error message itself, with
```
err.args += (f".....",)
raise err
```

Reply:
Adding more elements to the `.args` tuple ends up being shown like this:

```
astropy.units.core.UnitConversionError: ("'pc' (length) and 'deg' (angle) are not convertible", "Unit 'deg' (angle) could not be applied to 'distance'. ", 'This can occur when passing units for some coordinate components when other components are specified as Quantity objects. Either pass a list of units for all components (and unit-less coordinate data), or pass Quantities for all components')
```

which I find a tad ugly. I think I will raise a `UnitConversionError` but leave the custom message.

Reply:
Sure, the `raise ... from err` you had is fine too!